File: python/src/sde/engines/_write_fences.py
"""Native DDL and drain operations for the public write-fence protocol.

These objects need dedicated provisioning connections. They do not run on the application's write
path. Metadata is checked as well as created: an identically named, unrelated table is not ours.
"""

from __future__ import annotations

import re
from typing import Any

from ..errors import EngineError, MigrationRefused
from ..schema import QUOTE
from ..write_fence import DRAIN_TABLE, EPOCH_COLUMN, FenceMetadata, fence_state


class PostgresFences:
    def __init__(self, connection: Any) -> None:
        if not connection.autocommit:
            raise MigrationRefused("write-fence DDL needs a dedicated autocommit connection")
        self._cx = connection
        self._quote = QUOTE["postgres"]

    def _query(self, query: str, values: list[Any] | None = None) -> list[Any]:
        try:
            with self._cx.cursor() as cursor:
                cursor.execute(query, values)
                return list(cursor.fetchall()) if cursor.description else []
        except Exception as exc:
            raise EngineError(
                f"write-fence operation failed; inspect or resume its state: {exc}"
            ) from exc

    def _idle(self) -> None:
        if int(self._cx.info.transaction_status) != 0:
            raise MigrationRefused("write-fence DDL cannot run inside an application transaction")

    def metadata(self, table: str) -> FenceMetadata:
        quoted = self._quote(table)
        rows = self._query(
            "SELECT c.oid::text, c.relkind, EXISTS (SELECT 1 FROM pg_inherits i "
            "WHERE i.inhrelid=c.oid OR i.inhparent=c.oid) FROM pg_class c "
            "WHERE c.oid=to_regclass(%s)",
            [quoted],
        )
        if len(rows) != 1:
            raise EngineError("write-fence table does not exist")
        identity, kind, inherits = rows[0]
        if kind != "r" or inherits:
            raise MigrationRefused(
                "write fences support ordinary PostgreSQL tables without inheritance"
            )
        columns = self._query(
            "SELECT a.atttypid='bigint'::regtype, a.attnotnull, a.attgenerated, "
            "pg_get_expr(d.adbin,d.adrelid) FROM pg_attribute a "
            "LEFT JOIN pg_attrdef d ON d.adrelid=a.attrelid AND d.adnum=a.attnum "
            "WHERE a.attrelid=to_regclass(%s) AND a.attname=%s AND NOT a.attisdropped",
            [quoted, EPOCH_COLUMN],
        )
        valid = (
            bool(columns)
            and columns[0][:3] == (True, True, "")
            and columns[0][3]
            in (
                "0",
                "'0'::bigint",
            )
        )
        constraints = self._query(
            "SELECT conname, pg_get_constraintdef(oid) FROM pg_constraint "
            "WHERE conrelid=to_regclass(%s) AND contype='c'",
            [quoted],
        )
        return FenceMetadata(
            str(identity),
            "valid" if valid else "conflict" if columns else "absent",
            {str(name): str(expression) for name, expression in constraints},
        )

    def add_column(self, table: str) -> None:
        self._idle()
        self._query(
            f"ALTER TABLE {self._quote(table)} ADD COLUMN IF NOT EXISTS "
            f"{self._quote(EPOCH_COLUMN)} bigint NOT NULL DEFAULT 0"
        )

    def add_constraint(self, table: str, name: str, expression: str) -> None:
        self._idle()
        if name in self.metadata(table).constraints:
            return
        expression = {"1": "true", "0": "false"}.get(expression, expression)
        self._query(
            f"ALTER TABLE {self._quote(table)} ADD CONSTRAINT {self._quote(name)} "
            f"CHECK ({expression}) NOT VALID"
        )

    def drop_constraint(self, table: str, name: str) -> None:
        self._idle()
        self._query(
            f"ALTER TABLE {self._quote(table)} DROP CONSTRAINT IF EXISTS {self._quote(name)}"
        )

    def drain(self, table: str, *, project_id: str, hold: str) -> None:
        self._idle()
        # Even when a hold already existed, returning from this transaction proves that every
        # writer which held a conflicting table lock has completed. A retry repeats this proof.
        with self._cx.transaction():
            self._query(f"LOCK TABLE {self._quote(table)} IN SHARE ROW EXCLUSIVE MODE")
            _held(self.metadata(table), project_id, hold)

    def restore(self, table: str, *, project_id: str, hold: str) -> None:
        _held(self.metadata(table), project_id, hold)


class ClickHouseFences:
    def __init__(self, client: Any) -> None:
        self._cx = client
        self._quote = QUOTE["clickhouse"]

    def _query(self, query: str, parameters: dict[str, Any] | None = None) -> list[Any]:
        try:
            return list(self._cx.query(query, parameters=parameters or {}).result_rows)
        except Exception as exc:
            raise EngineError(f"write-fence metadata query failed: {exc}") from exc

    def _command(self, query: str) -> None:
        try:
            self._cx.command(query)
        except Exception as exc:
            raise EngineError(
                f"write-fence DDL failed; the table may remain closed or detached; "
                f"resume the same operation: {exc}"
            ) from exc

    def _table(self, table: str) -> list[Any]:
        return self._query(
            "SELECT toString(uuid), engine, formatQuery(create_table_query) FROM system.tables "
            "WHERE database=currentDatabase() AND name={table:String}",
            {"table": table},
        )

    def metadata(self, table: str) -> FenceMetadata:
        database = self._query("SELECT engine FROM system.databases WHERE name=currentDatabase()")
        if database != [("Atomic",)]:
            raise MigrationRefused("write fences require a local Atomic ClickHouse database")
        rows = self._table(table)
        if len(rows) != 1:
            raise EngineError(
                "write-fence table does not exist or is detached; resume the same operation"
            )
        identity, kind, create = rows[0]
        if kind not in ("MergeTree", "ReplacingMergeTree"):
            raise MigrationRefused(
                "write fences support local MergeTree and ReplacingMergeTree tables"
            )
        columns = self._query(
            "SELECT type, default_kind, default_expression FROM system.columns "
            "WHERE database=currentDatabase() AND table={table:String} AND name={column:String}",
            {"table": table, "column": EPOCH_COLUMN},
        )
        valid = columns == [("Int64", "DEFAULT", "0")]
        constraints = {
            name: expression.rstrip().removesuffix(",")
            for name, expression in re.findall(
                r'^\s*CONSTRAINT\s+[`"]?(__sde_f_[A-Za-z0-9_]+)[`"]?\s+CHECK\s+([^\n]+)',
                str(create),
                re.MULTILINE,
            )
        }
        return FenceMetadata(
            str(identity),
            "valid" if valid else "conflict" if columns else "absent",
            constraints,
        )

    def add_column(self, table: str) -> None:
        self._command(
            f"ALTER TABLE {self._quote(table)} ADD COLUMN IF NOT EXISTS "
            f"{self._quote(EPOCH_COLUMN)} Int64 DEFAULT 0"
        )

    def add_constraint(self, table: str, name: str, expression: str) -> None:
        self._command(
            f"ALTER TABLE {self._quote(table)} ADD CONSTRAINT IF NOT EXISTS "
            f"{self._quote(name)} CHECK {expression}"
        )

    def drop_constraint(self, table: str, name: str) -> None:
        self._command(
            f"ALTER TABLE {self._quote(table)} DROP CONSTRAINT IF EXISTS {self._quote(name)}"
        )

    def _drain_log(self) -> None:
        self._command(
            f"CREATE TABLE IF NOT EXISTS {self._quote(DRAIN_TABLE)} ("
            "table_name String, table_uuid UUID, project_id FixedString(32), hold String) "
            "ENGINE=MergeTree ORDER BY (table_uuid, project_id, hold)"
        )
        columns = self._query(
            "SELECT name,type FROM system.columns WHERE database=currentDatabase() "
            "AND table={table:String} ORDER BY name",
            {"table": DRAIN_TABLE},
        )
        if columns != [
            ("hold", "String"),
            ("project_id", "FixedString(32)"),
            ("table_name", "String"),
            ("table_uuid", "UUID"),
        ]:
            raise MigrationRefused("the reserved write-fence drain log has an incompatible schema")
        kinds = self._query(
            "SELECT engine FROM system.tables WHERE database=currentDatabase() "
            "AND name={table:String}",
            {"table": DRAIN_TABLE},
        )
        if kinds != [("MergeTree",)]:
            raise MigrationRefused("the reserved write-fence drain log has an incompatible engine")

    def drain(self, table: str, *, project_id: str, hold: str) -> None:
        metadata = self.metadata(table)
        _held(metadata, project_id, hold)
        self._drain_log()
        # Store the exact Atomic UUID before DETACH. A failed response can leave a permanently
        # detached table; resumption must not attach an unrelated table just because its name fits.
        try:
            self._cx.insert(
                DRAIN_TABLE,
                [[table, metadata.identity, project_id, hold]],
                column_names=["table_name", "table_uuid", "project_id", "hold"],
            )
        except Exception as exc:
            raise EngineError(f"write-fence drain intent was not confirmed: {exc}") from exc
        self._command(f"DETACH TABLE {self._quote(table)} PERMANENTLY SYNC")
        self._command(f"ATTACH TABLE {self._quote(table)}")
        after = self.metadata(table)
        if after.identity != metadata.identity:
            raise EngineError("the write-fence table identity changed while draining it")
        _held(after, project_id, hold)

    def restore(self, table: str, *, project_id: str, hold: str) -> None:
        if self._table(table):
            _held(self.metadata(table), project_id, hold)
            return
        # Reading a missing log fails closed. This path does not create one and must not guess
        # ownership from a detached table's name or from the supplied request alone.
        records = self._query(
            f"SELECT DISTINCT toString(table_uuid) FROM {self._quote(DRAIN_TABLE)} "
            "WHERE table_name={table:String} AND project_id={project:String} "
            "AND hold={hold:String}",
            {"table": table, "project": project_id, "hold": hold},
        )
        detached = self._query(
            "SELECT toString(uuid) FROM system.detached_tables "
            "WHERE database=currentDatabase() AND table={table:String} AND is_permanently=1",
            {"table": table},
        )
        if len(records) != 1 or detached != records:
            raise MigrationRefused("no matching durable intent for this detached write-fence table")
        self._command(f"ATTACH TABLE {self._quote(table)}")
        metadata = self.metadata(table)
        if metadata.identity != records[0][0]:
            raise EngineError("the restored write-fence table has another identity")
        _held(metadata, project_id, hold)


def _held(metadata: FenceMetadata, project_id: str, hold: str) -> None:
    state = fence_state(metadata)
    if state.project_id != project_id or hold not in state.holds:
        raise MigrationRefused("the requested project's write barrier is not installed")

File: typescript/src/engines/_write-fences.ts
/** Native fencing DDL for dedicated provisioning connections. No placement decision lives here. */
import { EngineError, MigrationRefused } from '../errors.js'
import { QUOTE } from '../schema.js'
import type { Row } from '../session.js'
import { DRAIN_TABLE, EPOCH_COLUMN, fenceState, type FenceBackend, type FenceMetadata } from '../write-fence.js'

interface PgTransport {
  query(sql: string, values?: readonly unknown[]): Promise<readonly Row[]>
  transaction(body: () => Promise<void>): Promise<void>
  busy(): boolean
}
interface ChTransport {
  query(sql: string): Promise<readonly Row[]>
  command(sql: string): Promise<void>
  insert(table: string, rows: readonly Row[]): Promise<void>
  literal(value: unknown): string
}

function held(metadata: FenceMetadata, projectId: string, hold: string): void {
  const state = fenceState(metadata)
  if (state.projectId !== projectId || !state.holds.includes(hold)) {
    throw new MigrationRefused("the requested project's write barrier is not installed")
  }
}

export class PostgresFences implements FenceBackend {
  private readonly quote = QUOTE['postgres'] as (name: string) => string
  constructor(private readonly transport: PgTransport) {}
  private idle(): void {
    if (this.transport.busy()) {
      throw new MigrationRefused('write-fence DDL cannot run inside an application transaction')
    }
  }
  async metadata(table: string): Promise<FenceMetadata> {
    const quoted = this.quote(table)
    const rows = await this.transport.query(
      'SELECT c.oid::text AS identity, c.relkind AS kind, EXISTS (SELECT 1 FROM pg_inherits i ' +
      'WHERE i.inhrelid=c.oid OR i.inhparent=c.oid) AS inherits FROM pg_class c ' +
      'WHERE c.oid=to_regclass($1)', [quoted],
    )
    if (rows.length !== 1) throw new EngineError('write-fence table does not exist')
    const row = rows[0] as Row
    if (row['kind'] !== 'r' || row['inherits'] !== false) {
      throw new MigrationRefused('write fences support ordinary PostgreSQL tables without inheritance')
    }
    const columns = await this.transport.query(
      "SELECT a.atttypid='bigint'::regtype AS typed, a.attnotnull AS required, " +
      'a.attgenerated AS generated, pg_get_expr(d.adbin,d.adrelid) AS expression FROM pg_attribute a ' +
      'LEFT JOIN pg_attrdef d ON d.adrelid=a.attrelid AND d.adnum=a.attnum ' +
      'WHERE a.attrelid=to_regclass($1) AND a.attname=$2 AND NOT a.attisdropped', [quoted, EPOCH_COLUMN],
    )
    const column = columns[0]
    const valid = column !== undefined && column['typed'] === true && column['required'] === true &&
      column['generated'] === '' && (column['expression'] === '0' || column['expression'] === "'0'::bigint")
    const constraints = await this.transport.query(
      "SELECT conname AS name, pg_get_constraintdef(oid) AS expression FROM pg_constraint " +
      "WHERE conrelid=to_regclass($1) AND contype='c'", [quoted],
    )
    return {
      identity: String(row['identity']), column: valid ? 'valid' : columns.length > 0 ? 'conflict' : 'absent',
      constraints: Object.fromEntries(constraints.map((item) => [String(item['name']), String(item['expression'])])),
    }
  }
  async addColumn(table: string): Promise<void> {
    this.idle()
    await this.transport.query(`ALTER TABLE ${this.quote(table)} ADD COLUMN IF NOT EXISTS ` +
      `${this.quote(EPOCH_COLUMN)} bigint NOT NULL DEFAULT 0`)
  }
  async addConstraint(table: string, name: string, expression: string): Promise<void> {
    this.idle()
    if (name in (await this.metadata(table)).constraints) return
    const predicate = expression === '1' ? 'true' : expression === '0' ? 'false' : expression
    await this.transport.query(`ALTER TABLE ${this.quote(table)} ADD CONSTRAINT ${this.quote(name)} ` +
      `CHECK (${predicate}) NOT VALID`)
  }
  async dropConstraint(table: string, name: string): Promise<void> {
    this.idle()
    await this.transport.query(`ALTER TABLE ${this.quote(table)} DROP CONSTRAINT IF EXISTS ${this.quote(name)}`)
  }
  async drain(table: string, options: { projectId: string; hold: string }): Promise<void> {
    this.idle()
    await this.transport.transaction(async () => {
      await this.transport.query(`LOCK TABLE ${this.quote(table)} IN SHARE ROW EXCLUSIVE MODE`)
      held(await this.metadata(table), options.projectId, options.hold)
    })
  }
  async restore(table: string, options: { projectId: string; hold: string }): Promise<void> {
    held(await this.metadata(table), options.projectId, options.hold)
  }
}

export class ClickHouseFences implements FenceBackend {
  private readonly quote = QUOTE['clickhouse'] as (name: string) => string
  constructor(private readonly transport: ChTransport) {}
  private table(table: string): Promise<readonly Row[]> {
    return this.transport.query('SELECT toString(uuid) AS identity, engine, ' +
      'formatQuery(create_table_query) AS definition FROM system.tables ' +
      'WHERE database=currentDatabase() AND name=' + this.transport.literal(table))
  }
  async metadata(table: string): Promise<FenceMetadata> {
    const databases = await this.transport.query('SELECT engine FROM system.databases WHERE name=currentDatabase()')
    if (databases.length !== 1 || databases[0]?.['engine'] !== 'Atomic') {
      throw new MigrationRefused('write fences require a local Atomic ClickHouse database')
    }
    const rows = await this.table(table)
    if (rows.length !== 1) {
      throw new EngineError('write-fence table does not exist or is detached; resume the same operation')
    }
    const row = rows[0] as Row
    if (row['engine'] !== 'MergeTree' && row['engine'] !== 'ReplacingMergeTree') {
      throw new MigrationRefused('write fences support local MergeTree and ReplacingMergeTree tables')
    }
    const columns = await this.transport.query('SELECT type, default_kind, default_expression FROM system.columns ' +
      'WHERE database=currentDatabase() AND table=' + this.transport.literal(table) +
      ' AND name=' + this.transport.literal(EPOCH_COLUMN))
    const column = columns[0]
    const valid = columns.length === 1 && column?.['type'] === 'Int64' &&
      column['default_kind'] === 'DEFAULT' && column['default_expression'] === '0'
    const constraints: Record<string, string> = {}
    const pattern = /^\s*CONSTRAINT\s+[`"]?(__sde_f_[A-Za-z0-9_]+)[`"]?\s+CHECK\s+([^\n]+)/gm
    for (const match of String(row['definition']).matchAll(pattern)) {
      constraints[match[1] as string] = (match[2] as string).trimEnd().replace(/,$/, '')
    }
    return {
      identity: String(row['identity']), column: valid ? 'valid' : columns.length > 0 ? 'conflict' : 'absent',
      constraints,
    }
  }
  async addColumn(table: string): Promise<void> {
    await this.transport.command(`ALTER TABLE ${this.quote(table)} ADD COLUMN IF NOT EXISTS ` +
      `${this.quote(EPOCH_COLUMN)} Int64 DEFAULT 0`)
  }
  async addConstraint(table: string, name: string, expression: string): Promise<void> {
    await this.transport.command(`ALTER TABLE ${this.quote(table)} ADD CONSTRAINT IF NOT EXISTS ` +
      `${this.quote(name)} CHECK ${expression}`)
  }
  async dropConstraint(table: string, name: string): Promise<void> {
    await this.transport.command(`ALTER TABLE ${this.quote(table)} DROP CONSTRAINT IF EXISTS ${this.quote(name)}`)
  }
  private async drainLog(): Promise<void> {
    await this.transport.command(`CREATE TABLE IF NOT EXISTS ${this.quote(DRAIN_TABLE)} (` +
      'table_name String, table_uuid UUID, project_id FixedString(32), hold String) ' +
      'ENGINE=MergeTree ORDER BY (table_uuid, project_id, hold)')
    const columns = await this.transport.query('SELECT name,type FROM system.columns WHERE database=currentDatabase() ' +
      'AND table=' + this.transport.literal(DRAIN_TABLE) + ' ORDER BY name')
    const actual = columns.map((row) => [row['name'], row['type']])
    const expected = [['hold', 'String'], ['project_id', 'FixedString(32)'], ['table_name', 'String'], ['table_uuid', 'UUID']]
    if (JSON.stringify(actual) !== JSON.stringify(expected)) {
      throw new MigrationRefused('the reserved write-fence drain log has an incompatible schema')
    }
    const kinds = await this.transport.query('SELECT engine FROM system.tables WHERE database=currentDatabase() AND name=' +
      this.transport.literal(DRAIN_TABLE))
    if (kinds.length !== 1 || kinds[0]?.['engine'] !== 'MergeTree') {
      throw new MigrationRefused('the reserved write-fence drain log has an incompatible engine')
    }
  }
  async drain(table: string, options: { projectId: string; hold: string }): Promise<void> {
    const metadata = await this.metadata(table)
    held(metadata, options.projectId, options.hold)
    await this.drainLog()
    await this.transport.insert(DRAIN_TABLE, [{ table_name: table, table_uuid: metadata.identity,
      project_id: options.projectId, hold: options.hold }])
    await this.transport.command(`DETACH TABLE ${this.quote(table)} PERMANENTLY SYNC`)
    await this.transport.command(`ATTACH TABLE ${this.quote(table)}`)
    const after = await this.metadata(table)
    if (after.identity !== metadata.identity) throw new EngineError('the write-fence table identity changed while draining it')
    held(after, options.projectId, options.hold)
  }
  async restore(table: string, options: { projectId: string; hold: string }): Promise<void> {
    if ((await this.table(table)).length > 0) {
      held(await this.metadata(table), options.projectId, options.hold)
      return
    }
    const literal = this.transport.literal
    const records = await this.transport.query(`SELECT DISTINCT toString(table_uuid) AS identity FROM ${this.quote(DRAIN_TABLE)} ` +
      `WHERE table_name=${literal(table)} AND project_id=${literal(options.projectId)} AND hold=${literal(options.hold)}`)
    const detached = await this.transport.query('SELECT toString(uuid) AS identity FROM system.detached_tables ' +
      `WHERE database=currentDatabase() AND table=${literal(table)} AND is_permanently=1`)
    if (records.length !== 1 || detached.length !== 1 || records[0]?.['identity'] !== detached[0]?.['identity']) {
      throw new MigrationRefused('no matching durable intent for this detached write-fence table')
    }
    await this.transport.command(`ATTACH TABLE ${this.quote(table)}`)
    const metadata = await this.metadata(table)
    if (metadata.identity !== records[0]?.['identity']) throw new EngineError('the restored write-fence table has another identity')
    held(metadata, options.projectId, options.hold)
  }
}


export async function fenceIO<T>(operation: () => Promise<T>): Promise<T> {
  try {
    return await operation()
  } catch (error) {
    if (error instanceof EngineError || error instanceof MigrationRefused) throw error
    const detail = error instanceof Error ? error.message : String(error)
    throw new EngineError('write-fence operation failed; inspect or resume its state: ' + detail)
  }
}
